### PR TITLE
Enable unit support on all functional models

### PR DESCRIPTION
**Pull request by astrofrog**
The goal of this PR is to add tests for all functional models with units and fix any issues that come up. This is now ready for review. The only thing is that the absorption gaussian can't be made unit-aware because of the ``1 - `` term - there is no way to attach a unit to ``1`` (I personally find this model strange since really one should do Const1D - Gaussian1D). Maybe we should replace ``1`` by a model parameter, but I suggest deferring that to a separate PR (I'm trying to keep my modeling PRs from getting out of control...)

### CODE REVIEW

**Comment by bsipocz** [comment]
@astrofrog - I have manually cancelled the last travis, sadly the skipping is not working as expected in PRs, but I'm on it.

**Comment by pllim** [comment]
@nden asked if you plan to address non-functional models, like projections, mapping, etc.

**Comment by astrofrog** [comment]
@pllim - I think projections are currently tested already thanks to @Cadair, but haven't looked into mappings yet. I do plan to take a look at this though.

**Comment by astrofrog** [comment]
@nden @pllim - can you review this when you get a chance? Thanks!

File: astropy/modeling/functional_models.py [hunk 1]
```diff
@@ -714,20 +749,33 @@ class Sine1D(Fittable1DModel):
     @staticmethod
     def evaluate(x, amplitude, frequency, phase):
         """One dimensional Sine model function"""
-
-        return amplitude * np.sin(2 * np.pi * frequency * x + 2 * np.pi * phase)
+        argument = TWOPI * (frequency * x + phase)
+        if isinstance(argument, Quantity):
```
Comment (pllim) [inline astropy/modeling/functional_models.py]: If it is not Quantity and you want to attach radian unit to it, I get it. But here, you are multiplying radian unit to it but only if it is a Quantity, which I do not get. What are you trying to do here?

File: astropy/modeling/tests/test_functional_models_quantities.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+                                 MexicanHat2D, AiryDisk2D, Moffat2D, Sersic2D)
+from ..fitting import LevMarLSQFitter
+
+try:
+    from scipy import optimize
+    HAS_SCIPY = True
+except ImportError:
+    HAS_SCIPY = False
+
+# TODO: GaussianAbsorption1D doesn't work with units because the 1- part doesn't
```
Comment (pllim) [inline astropy/modeling/tests/test_functional_models_quantities.py]: Good question. `GaussianAbsorption1D` as it is now is only useful as a *multiplicative* factor to a continuum to create an absorption line. Otherwise, you can create absorption line without it by using `continuum_model - Gaussian1D`. (This model might had been written a long time ago before the math operators worked properly?) I wonder if `GaussianAbsorption1D` is used by anyone at all. Maybe deprecate it for future removal? That way, we can just say "this model is deprecated and does not support units".

Related reading: http://synphot.readthedocs.io/en/latest/synphot/spectrum.html#gaussian-absorption

File: astropy/modeling/tests/test_functional_models_quantities.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+        x = np.linspace(1, 3, 100) * model['evaluation'][0][0].unit
+        y = np.exp(-x.value ** 2) * model['evaluation'][0][1].unit
+        args = [x, y]
+    else:
+        x = np.linspace(1, 3, 100) * model['evaluation'][0][0].unit
+        y = np.linspace(1, 3, 100) * model['evaluation'][0][1].unit
+        z = np.exp(-x.value**2 - y.value**2) * model['evaluation'][0][2].unit
+        args = [x, y, z]
+    fitter = LevMarLSQFitter()
+    m_new = fitter(m, *args)
```
Comment (pllim) [inline astropy/modeling/tests/test_functional_models_quantities.py]: This simply tests that fitter does not crash but does not test whether the fitted values are actually correct (both values and units).

File: astropy/modeling/tests/test_functional_models_quantities.py [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+        else:
+            x, y, z = args
+            x_arr = u.Quantity([x, x])
+            y_arr = u.Quantity([y, y])
+            result = m(x_arr, y_arr)
+            assert_quantity_allclose(result, u.Quantity([z, z]))
+
+
+@pytest.mark.parametrize('model', MODELS)
+def test_models_bounding_box(model):
```
Comment (pllim) [inline astropy/modeling/tests/test_functional_models_quantities.py]: Why is there a need to test this as part of this PR? Bounding box seems unrelated here. If your goal is to test whether bounding box has the correct unit, that does not seem to be tested.

File: astropy/modeling/core.py [hunk 5]
```diff
@@ -1475,22 +1475,17 @@ def prepare_inputs(self, *inputs, **kwargs):
             for i in range(len(inputs)):
 
                 input_name = self.inputs[i]
+                input_unit = self.input_units.get(input_name, None)
```
Comment (nden) [inline astropy/modeling/core.py]: At this point it's always a dict. The check for None is done earlier.

File: astropy/modeling/functional_models.py (hunk 1 above)
Comment (nden) [inline astropy/modeling/functional_models.py]: I wonder about this too. The code works but I'm not sure I understand why it's done this way.

**Comment by nden** [comment]
@astrofrog This looks good to me. 
I got one error  in GWCS due to a rotation model initialized with a SkyCoord object.
If no one is working on rotations I can try to add units there.

File: astropy/modeling/tests/test_functional_models_quantities.py (hunk 3 above)
Comment (astrofrog) [inline astropy/modeling/tests/test_functional_models_quantities.py]: That is correct - I should at least make sure that the resulting models have units - however I think checking the fitted parameter values is beyond the scopes of these tests

File: astropy/modeling/functional_models.py (hunk 1 above)
Comment (astrofrog) [inline astropy/modeling/functional_models.py]: Ah yes - I admit it's a little confusing - the reason for this is because if frequency and x are quantities, argument will end up being dimensionless because TWOPI doesn't have a unit (even though it should be radians). But then np.sin(dimensionless) crashes.

Another option would be to do ``argument = argument.value`` which would work too. I might do that instead since that will work too, and will put a comment.

File: astropy/modeling/tests/test_functional_models_quantities.py (hunk 2 above)
Comment (astrofrog) [inline astropy/modeling/tests/test_functional_models_quantities.py]: Maybe we could first ask if anyone is using it on astropy-dev - another option is to make the continuum level a parameter.

File: astropy/modeling/tests/test_functional_models_quantities.py (hunk 4 above)
Comment (astrofrog) [inline astropy/modeling/tests/test_functional_models_quantities.py]: Because for some models, bounding_box crashed if units were present, so I had to fix those and wanted a regression test. I agree we could check the actual values, but the purpose here is first and foremost to check that it does anything at all.

**Comment by astrofrog** [comment]
@nden - if you'd like to work on rotations that would be great! Just ping me if you run into any issues.

File: astropy/modeling/tests/test_functional_models_quantities.py (hunk 3 above)
Comment (astrofrog) [inline astropy/modeling/tests/test_functional_models_quantities.py]: Ok, I've now expanded that to make sure the resulting model has units and that they are compatible with the starting units.

File: astropy/modeling/tests/test_functional_models_quantities.py (hunk 4 above)
Comment (astrofrog) [inline astropy/modeling/tests/test_functional_models_quantities.py]: I've updated this to now check for the actual accuracy of the bounding box values too

File: astropy/modeling/tests/test_functional_models_quantities.py (hunk 2 above)
Comment (astrofrog) [inline astropy/modeling/tests/test_functional_models_quantities.py]: I opened an issue to discuss this: https://github.com/astropy/astropy/issues/6195

**Comment by astrofrog** [comment]
@nden @pllim - thanks for the reviews! I think I've addressed all your concerns/comments. If the CI passes, i think this is good to go, but I'll let one of you merge if you are happy with it (let's wait for Travis etc to run first though).

**Comment by astrofrog** [comment]
@nden - just for info I plan to look into power laws and polynomials, but I'll leave rotations for you to take a look at.

File: astropy/modeling/tests/test_functional_models_quantities.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+        pytest.skip()
+
+    m = model['class'](**model['parameters'])
+
+    if model['bounding_box']:
+        # A bounding box may have inhomogeneous units so we need to check the
+        # values one by one.
+        for i in range(len(model['bounding_box'])):
+            bbox = m.bounding_box
+            print(bbox)
```
Comment (pllim) [inline astropy/modeling/tests/test_functional_models_quantities.py]: print function does not seem necessary. Is this leftover from debugging phase?

File: astropy/modeling/tests/test_functional_models_quantities.py [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+ 'evaluation':[(3 * u.arcsec, 2.5 * u.arcsec, 2.829990489 * u.MJy/u.sr)],
+ 'bounding_box': False},
+]
+
+MODELS = MODELS_1D + MODELS_2D
+
+SCIPY_MODELS = set([Sersic1D, Sersic2D, AiryDisk2D])
+
+@pytest.mark.parametrize('model', MODELS)
+def test_models_evaluatate_without_units(model):
```
Comment (pllim) [inline astropy/modeling/tests/test_functional_models_quantities.py]: evaluatate -> evaluate (multiple occurrences)

**Comment by astrofrog** [comment]
@pllim - I've implemented the fixes. In general I think we should only ask people to squash if they spend say 10 commits making trivial changes, or if they accidentally add then remove a large file, but for 600+ line changes, I don't think it's unreasonable to have a dozen commits.
